core/src/types/buffer.rs

```diff
@@ -176,6 +176,27 @@ impl Buffer {
         self.len() == 0
     }
 
+    /// Number of [`Bytes`] in [`Buffer`].
+    ///
+    /// For contiguous buffer, it's always 1. For non-contiguous buffer, it's number of bytes
+    /// available for use.
+    pub fn count(&self) -> usize {
+        match &self.0 {
+            Inner::Contiguous(_) => 1,
+            Inner::NonContiguous { parts, idx, .. } => parts.len() - idx,
+        }
+    }
+
+    /// Get current [`Bytes`].
+    pub fn current(&self) -> Bytes {
+        match &self.0 {
+            Inner::Contiguous(inner) => inner.clone(),
+            Inner::NonContiguous {
+                parts, idx, offset, ..
+            } => parts[*idx].slice(*offset..),
```

Comment:
Do we need to perform more checks like we do in `Buf::chunk`?

Reply:
If we cannot assume offset + size will be exactly at the end of all buffer, then yes, we should check. Is that the design?

Reply:
> If we cannot assume offset + size will be exactly at the end of all buffer,

`size` is the `Buffer`'s size. So `offset + size` is not the end of all buffer. It's possible that `parts[*idx].slice(*offset..).len() < size`.